Data corruption exception in System tests.

**Problem description**
In some system test runs such as ReadTxnWriteScaleWithFailoverTest, when bookies are repeatedly failing, a `DataCorruptionException` occurs. Example:

```
segmentstore_2019-03-14.59.log:2019-03-14 06:25:48,526 1412251 [core-18] INFO  i.p.s.s.h.ZKSegmentContainerMonitor - Container Changes: Desired = [0, 1, 2, 3], Current = [0, 1, 2, 3], PendingTasks = [], ToStart = [], ToStop = [].
2019-03-14 06:25:52,893 1416618 [core-25] WARN  i.p.s.s.c.StreamSegmentContainer - SegmentContainer[1]: Writer failed. Shutting down StreamSegmentContainer.
java.util.concurrent.CompletionException: io.pravega.segmentstore.server.DataCorruptionException: Unable to retrieve CacheContents for 'AggregatedAppend: SegmentId = 1, Offsets = [892517-2320514), Se
qNo = 18494'.
        at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292)
  Caused by: io.pravega.segmentstore.server.DataCorruptionException: Unable to retrieve CacheContents for 'AggregatedAppend: SegmentId = 1, Offsets = [892517-2320514), SeqNo = 18494'.
        at io.pravega.segmentstore.server.writer.SegmentAggregator.getFlushArgs(SegmentAggregator.java:826)
        at io.pravega.segmentstore.server.writer.SegmentAggregator.flushPendingAppends(SegmentAggregator.java:754)
        at io.pravega.segmentstore.server.writer.SegmentAggregator.lambda$flushExcess$14(SegmentAggregator.java:696)
        ... 11 common frames omitted
        
segmentstore_2019-03-14.60.log:2019-03-14 06:25:58,526 1422251 [core-24] INFO  i.p.s.s.h.ZKSegmentContainerMonitor - Container Changes: Desired = [0, 1, 2, 3], Current = [0, 2, 3], PendingTasks = [], ToStart = [1], ToStop = [].
```

--- a/segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/WriterTableProcessor.java
+++ b/segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/WriterTableProcessor.java
@@ -176,6 +176,17 @@ public class WriterTableProcessor implements WriterSegmentProcessor {
                 });
     }
 
+    @Override
+    public String toString() {
+        return String.format(
+                "[%d: %s] Count = %d, LastOffset = %s, LUSN = %d",
+                this.connector.getMetadata().getId(),
+                this.connector.getMetadata().getName(),
+                this.aggregator.size(),
+                this.lastAddedOffset,
+                getLowestUncommittedSequenceNumber());
+    }
+
     //endregion
 
     //region Helpers
@@ -469,6 +480,10 @@ public class WriterTableProcessor implements WriterSegmentProcessor {
             }
 
             return true;
+        }
+
+        synchronized int size() {
+            return this.appendOffsets.size();
         }
 
         @Override

--- a/segmentstore/server/src/main/java/io/pravega/segmentstore/server/writer/AckCalculator.java
+++ b/segmentstore/server/src/main/java/io/pravega/segmentstore/server/writer/AckCalculator.java
@@ -11,6 +11,7 @@ package io.pravega.segmentstore.server.writer;
 
 import com.google.common.base.Preconditions;
 import io.pravega.segmentstore.server.WriterSegmentProcessor;
+import io.pravega.segmentstore.server.logs.operations.Operation;
 
 /**
  * Helper class that calculates Ack Sequence Numbers based on input information from OperationProcessors.
@@ -31,9 +32,10 @@ class AckCalculator {
     /**
      * Determines the largest Sequence Number that can be safely truncated from the Writer's Data Source. All operations
      * up to, and including the one for this Sequence Number have been successfully committed to External Storage.
-     * <p>
+     *
      * The Sequence Number we acknowledge has the property that all operations up to, and including it, have been
      * committed to Storage.
+     *
      * This can only be calculated by looking at all the active SegmentAggregators and picking the Lowest Uncommitted
      * Sequence Number (LUSN) among all of those Aggregators that have any outstanding data. The LUSN for each aggregator
      * has the property that, within the context of that Aggregator alone, all Operations that have a Sequence Number (SN)
@@ -41,10 +43,16 @@ class AckCalculator {
      * all the active SegmentAggregators will give us the highest SN that can be safely truncated out of the OperationLog.
      * Note that LUSN still points to an uncommitted Operation, so we need to subtract 1 from it to obtain the highest SN
      * that can be truncated up to (and including).
+     *
      * If we have no active Aggregators, then we have committed all operations that were passed to us, so we can
      * safely truncate up to LastReadSequenceNumber.
      *
-     * @param processors The Processors to inspect for commit status.
+     * As opposed from {@link #getLowestUncommittedSequenceNumber}, this method should be called for
+     * {@link WriterSegmentProcessor} instances that deal with different Segments.
+     *
+     * @param processors The {@link WriterSegmentProcessor} to inspect for commit status.
+     * @param <T> {@link WriterSegmentProcessor} type.
+     * @return The Highest Committed Sequence Number.
      */
     <T extends WriterSegmentProcessor> long getHighestCommittedSequenceNumber(Iterable<T> processors) {
         long lowestUncommittedSeqNo = Long.MAX_VALUE;
@@ -61,5 +69,32 @@ class AckCalculator {
 
         lowestUncommittedSeqNo = Math.min(lowestUncommittedSeqNo, this.state.getLastReadSequenceNumber());
         return lowestUncommittedSeqNo;
+    }
+
+    /**
+     * Determines the lowest Sequence Number across the given {@link WriterSegmentProcessor} instances that has not
+     * yet been committed to Storage.
+     * <p>
+     * As opposed from {@link #getHighestCommittedSequenceNumber}, this method should be called for
+     * {@link WriterSegmentProcessor} instances that deal with the same Segment.
+     *
+     * @param processors The {@link WriterSegmentProcessor} to inspect.
+     * @param <T>        {@link WriterSegmentProcessor} type.
+     * @return The Lowest Uncommited Sequence Number.
+     */
+    <T extends WriterSegmentProcessor> long getLowestUncommittedSequenceNumber(Iterable<T> processors) {
+        // We need to find the lowest value across all processors that have uncommitted data.
+        long result = Long.MAX_VALUE;
+        for (WriterSegmentProcessor p : processors) {
+            // If a processor has uncommitted data, its LUSN will be a positive number. Otherwise it means it is up-to-date.
+            long sn = p.getLowestUncommittedSequenceNumber();
+            if (sn >= 0 && sn < result) {
+                result = sn;
+            }
+        }
+
+        // If at least one processor has uncommitted data, return the smallest LUSN among them. Otherwise report that
+        // everything is up-to date.
+        return result == Long.MAX_VALUE ? Operation.NO_SEQUENCE_NUMBER : result;
     }
 }

--- a/segmentstore/server/src/main/java/io/pravega/segmentstore/server/writer/SegmentAggregator.java
+++ b/segmentstore/server/src/main/java/io/pravega/segmentstore/server/writer/SegmentAggregator.java
@@ -228,7 +228,7 @@ class SegmentAggregator implements WriterSegmentProcessor, AutoCloseable {
     @Override
     public String toString() {
         return String.format(
-                "[%d: %s] Count = %d, LastOffset = %s, LUSN = %d LastFlush = %ds",
+                "[%d: %s] Count = %d, LastOffset = %s, LUSN = %d, LastFlush = %ds",
                 this.metadata.getId(),
                 this.metadata.getName(),
                 this.operations.size(),

--- a/segmentstore/server/src/main/java/io/pravega/segmentstore/server/writer/StorageWriter.java
+++ b/segmentstore/server/src/main/java/io/pravega/segmentstore/server/writer/StorageWriter.java
@@ -543,7 +543,7 @@ class StorageWriter extends AbstractThreadPoolService implements Writer {
     /**
      * Wraps a collection of WriterSegmentProcessors, including the main Segment Aggregator.
      */
-    private static class ProcessorCollection implements WriterSegmentProcessor {
+    private class ProcessorCollection implements WriterSegmentProcessor {
         private final SegmentAggregator aggregator;
         private final List<WriterSegmentProcessor> processors;
 
@@ -594,11 +594,9 @@ class StorageWriter extends AbstractThreadPoolService implements Writer {
 
         @Override
         public long getLowestUncommittedSequenceNumber() {
-            // We collect the lowest value across all processors.
-            return this.processors.stream()
-                                  .mapToLong(WriterSegmentProcessor::getLowestUncommittedSequenceNumber)
-                                  .min()
-                                  .orElse(Operation.NO_SEQUENCE_NUMBER);
+            return this.processors.size() == 1
+                    ? this.processors.get(0).getLowestUncommittedSequenceNumber()
+                    : StorageWriter.this.ackCalculator.getLowestUncommittedSequenceNumber(this.processors);
         }
 
         @Override

--- a/segmentstore/server/src/test/java/io/pravega/segmentstore/server/writer/AckCalculatorTests.java
+++ b/segmentstore/server/src/test/java/io/pravega/segmentstore/server/writer/AckCalculatorTests.java
@@ -13,6 +13,7 @@ import io.pravega.common.hash.RandomFactory;
 import io.pravega.segmentstore.server.SegmentOperation;
 import io.pravega.segmentstore.server.WriterFlushResult;
 import io.pravega.segmentstore.server.WriterSegmentProcessor;
+import io.pravega.segmentstore.server.logs.operations.Operation;
 import java.time.Duration;
 import java.util.ArrayList;
 import java.util.Random;
@@ -92,6 +93,36 @@ public class AckCalculatorTests {
         result = calc.getHighestCommittedSequenceNumber(processors);
         expectedResult = processors.stream().mapToLong(TestProcessor::getLowestUncommittedSequenceNumber).filter(sn -> sn >= 0).min().getAsLong() - 1;
         Assert.assertEquals("Unexpected result for Set with partial values when LRSN is infinite.", expectedResult, result);
+    }
+
+    /**
+     * Tests the {@link AckCalculator#getLowestUncommittedSequenceNumber} method.
+     */
+    @Test
+    public void testGetLowestUncommittedSequenceNumber() {
+        final int processorCount = 2;
+        WriterState state = new WriterState();
+        AckCalculator calc = new AckCalculator(state);
+
+        ArrayList<TestProcessor> processors = new ArrayList<>();
+        for (int i = 0; i < processorCount; i++) {
+            processors.add(new TestProcessor());
+        }
+
+        // Everything up-to-date.
+        processors.forEach(p -> p.setLowestUncommittedSequenceNumber(Operation.NO_SEQUENCE_NUMBER));
+        long lusn = calc.getLowestUncommittedSequenceNumber(processors);
+        Assert.assertEquals("Unexpected result when all processors up-to-date.", Operation.NO_SEQUENCE_NUMBER, lusn);
+
+        // One processor not up-to-date.
+        processors.get(0).setLowestUncommittedSequenceNumber(10);
+        lusn = calc.getLowestUncommittedSequenceNumber(processors);
+        Assert.assertEquals("Unexpected result when only one processor is up-to-date.", 10, lusn);
+
+        // Neither processor is up-to-date.
+        processors.get(1).setLowestUncommittedSequenceNumber(8);
+        lusn = calc.getLowestUncommittedSequenceNumber(processors);
+        Assert.assertEquals("Unexpected result when neither processor is up-to-date.", 8, lusn);
     }
 
     private static class TestProcessor implements WriterSegmentProcessor {

